### ISSUE TITLE
[Question] props - usage with arrays and other non-primitive types, best practice?

### ISSUE THREAD

**Issue by prof7bit** [WebReflection/uce#11]
I am still just playing around with this, so far liking it very much. As a relatively inexperienced JS developer I just got bitten by the following (it took an hour until it dawned on me what was going on)

Consider the following:

```javascript
define('my-app', {
    props: {
        foo: 42,
        bar: 23,
        items: []
    },

    // [...]
```

This worked just fine until I had the glorious idea of putting 2 instances of my-app into my html. The foo and bar properties were independent but the items array was shared between all instances.

I figured out this is because the array will be instantiated once during the call of define() already and then later a reference to it will be shared by all instances.

The only way around I have found is by creating it in the init() method. But should I then assign the array to this.props or only to this? Should i put an empty placeholder with the same name into props?

Or should I use props only for primitive types, because I also noticed it will not detect array manipulation via index or push, etc. and not trigger a render() anyway. My conclusion so far is I should not use props for any non-primitive types and live with that. Is this correct?

**Comment by WebReflection** [comment]
This is a valuable feedback, and arguably a library limitation or, to be honest, footgun.
I'll try to make it work the way you expected, thanks for the report.

**Comment by WebReflection** [comment]
I think your code should work as expected now ... it's a simple patch around primitives vs objects, so that arrays or objects are de/serialized to be unique per instance ... I hope this works well enough, as I didn't want to bloat the whole library with a deep clone that arguably is overkill for most common use cases.

**Comment by prof7bit** [comment]
Wow, this was fast! I just tested it and can confirm that it works now. 

---

The more I play around with your µhtml family of libraries and especially µce the more I like it. I wish I would have had something like this a couple of years ago when I was in the need of making a small web interface for an embedded Linux device. I had only a few weeks to make a decision, checked out only the big frameworks and went for vue.js at that time. It worked for what I needed it (visualize and control in realtime the state of n identical machines connected to a control box which itself had a websocket API streaming the state changes of all connected devices) but the whole experience (node dependencies, transpiling, etc) left me with very mixed feelings. On some days I spent more time wrestling with the framework than coding the application.

I now realize that today I could easily build the whole thing again with µce, no messing around with npm at all, just a copy of minified uce.js on my webserver and all my code in plain readable JavaScript with almost zero boilerplate to plug all components together and I would probably get it all done in less than half of the time. All features I would need seem to be present in uce.

So I take my hat off to you for creating such a powerful little piece of code that has the potential to make web development finally fun again!
